NETWORK_PROTOCOL.md

```diff
@@ -1 +1,110 @@
+
+After a subscription is closed, its Id can be reused, to prevent Ids from growing without limit. The subscriber MAY reuse an Id in a `subscribe` message after it has sent `cancel` or received `onComplete` or `onError` for that Id. If it does so, it MUST guarantee that the publisher will not receive messages meant for the previous subscription with that Id after it receives the second `subscribe` message.
+
+## Packed messaging
+
+In typical use, the most common messages by far are `onNext`. The overhead per message is typically 1 byte (message code) +  1-2 bytes (subscriber id) + 1-3 bytes (payload length) = 3-6 bytes total. When the message type is very small (e.g. an int), the overhead can be 100% or more.
+
+To reduce the overhead, the publisher can optionally declare that all stream elements will have a fixed size by setting the `onSubscribe.elementSize` field to a value greater than zero:
+
+    <-- onSubscribe(subscriber: Id, elementSize: varint)
+
+The publisher can then send not just `onNext` messages but also `onNextPacked` messages:
+
+    <-- onNextPacked(subscriber: Id, count: varint, messages: count * elementSize bytes)
+    
+Packing does not help if new data becomes available very frequently and must not be delayed before being sent. A typical example is a ticker source. It also can't be done if the client doesn't provide enough demand.
+
+## Split messaging
+
+A very large `onNext` message might significantly delay messages from other streams. Therefore, large stream elements can be optionally split across multiple messages. Publishers MAY split elements; subscribers MUST support this.
+
+When an element is split, the publisher will send one or more `onNextPart` messages, followed by a single `onNextLastPart`:
+
+    <-- onNextPart(subscriber: Id, element: Id, data: bytes)
+    <-- onNextLastPart(subscriber: Id, element: Id, data: bytes)
+
+`element` is an Id assigned by the Publisher; messages with the same `element` value, in the same stream, will be joined by the Subscriber. The order of the parts is that in which they were sent and received (the transport is required to provide ordered delivery).
+
+The subscriber's driver will typically join the parts transparently and deliver a single message to the application.
+
+## Closing the connection
+
+When the underlying transport is closed, both sides should release all related resources. This protocol version does not specify reusing previously negotiated state after reconnecting.
+
+The orderly way of closing a connection is to send a `goodbye` message, wait for acknowledgement and then close the underlying connection:
+
+    --> goodbye(reason: String)
+    <-- goodbye(reason: String)
+    
+Sending `goodbye` implicitly closes all open streams, equivalently to receiving `cancel` or `onError` messages.
```

Comment:
Might be good to use ACK for the acknowledgement instead. That way it is differentiated from the goodbye.